src/geode/mesh/core/solid_mesh.cpp

```diff
@@ -713,24 +713,86 @@ namespace geode
     }
 
     template < index_t dimension >
-    template < index_t T >
-    typename std::enable_if< T == 3, Vector3D >::type
-        SolidMesh< dimension >::polyhedron_facet_normal(
+    double SolidMesh< dimension >::polyhedron_volume(
+        index_t polyhedron_id ) const
+    {
+        if( nb_polyhedron_vertices( polyhedron_id ) < 4 )
+        {
+            return 0;
+        }
+        double volume{ 0 };
+        const auto first_pt_index = polyhedron_vertex( { polyhedron_id, 0 } );
+        const auto& p0 = point( first_pt_index );
+        const auto facets_vertices =
+            polyhedron_facets_vertices( polyhedron_id );
+        for( const auto facet_id : LIndices{ facets_vertices } )
+        {
+            const auto& facet_vertices = facets_vertices[facet_id];
+            if( absl::c_any_of(
+                    facet_vertices, [first_pt_index]( index_t vertex_id ) {
+                        return vertex_id == first_pt_index;
+                    } ) )
+            {
+                continue;
+            }
+            for( const auto i : LRange{ facet_vertices.size() - 2 } )

```

Comment:
```suggestion
            for( const auto i : LRange{ 2, facet_vertices.size() } )

```
Maybe this way, in case there is a degenerated facet?